### PR TITLE
Plane: support is_landing and is_taking_off in lua scripts

### PR DESCRIPTION
**Pull request by flatherz**
Adds is_landing and is_taking_off functionality for Plane as requested. Closes [23901](https://github.com/ArduPilot/ardupilot/issues/23901)

### CODE REVIEW

File: ArduPlane/mode_auto.cpp
Comment (magicrub) [inline ArduPlane/mode_auto.cpp:170]: the default is not necessary, can you remove it?

File: ArduPlane/mode_auto.cpp
Comment (magicrub) [inline ArduPlane/mode_auto.cpp:184]: no need for a default here either

File: ArduPlane/Plane.cpp
Comment (magicrub) [inline ArduPlane/Plane.cpp:47]: with quadplane enabled, this function has a double return which is unreachable code so please fix that. Also, watch out for whitespace on the # if, make it match with the rest of the code. This also needs to be moved to Arduplane.cpp under set_land_descent_rate() because there's already a group of functions there wrapped in AP_SCRIPTING_ENABLED

```suggestion
#if HAL_QUADPLANE_ENABLED
    if (plane.quadplane.in_vtol_land_descent()) {
        return true;
    }
#endif

    return control_mode->is_landing();
```

File: ArduPlane/Plane.cpp
Comment (magicrub) [inline ArduPlane/Plane.cpp:59]: with quadplane enabled, this function has a double return which is unreachable code so please fix that. Also, watch out for whitespace on the # if, make it match with the rest of the code. This also needs to be moved to Arduplane.cpp under set_land_descent_rate() because there's already a group of functions there wrapped in AP_SCRIPTING_ENABLED

```suggestion
#if HAL_QUADPLANE_ENABLED
    if (plane.quadplane.in_vtol_takeoff()) {
        return true;
    }
#endif

    return control_mode->is_taking_off();
```

File: ArduPlane/Plane.h
Comment (magicrub) [inline ArduPlane/Plane.h:526]: These belong down at the bottom next to the other public functions

**Review by magicrub — CHANGES_REQUESTED** [review]
great start! I have a couple of suggestions for improvements but otherwise I like it!

File: ArduPlane/mode_auto.cpp
Comment (magicrub) [inline ArduPlane/mode_auto.cpp:173]:
```suggestion
    return plane.mission.get_in_landing_sequence_flag();
```

File: ArduPlane/Plane.cpp
Comment (magicrub) [reply ArduPlane/Plane.cpp:59]:
```suggestion
    return control_mode->is_taking_off() || (started_flying_ms > 0 && millis() - started_flying_ms < 30000);
```

**Comment by magicrub** [comment]
@flatherz if you'd like, I can help out and make these changes for you.

**Comment by magicrub** [comment]
I also see that you have 5 commits and some are merges. Merges are not allowed in this repo. Please rebase onto master.

**Comment by flatherz** [comment]
> @flatherz if you'd like, I can help out and make these changes for you.

Thanks for the review! I've been on the road so haven't been able to take a look. Hoping to have some time over the next few days. I'll be in touch if I have any questions or need help!

File: ArduPlane/mode_auto.cpp
Comment (flatherz) [reply ArduPlane/mode_auto.cpp:170]: Sure thing! Working on a commit to remove this now.

File: ArduPlane/mode_auto.cpp
Comment (flatherz) [reply ArduPlane/mode_auto.cpp:184]: Same here.

File: ArduPlane/Plane.cpp
Comment (timtuxworth) [reply ArduPlane/Plane.cpp:59]: >     return control_mode->is_taking_off() || (started_flying_ms > 0 && millis() - started_flying_ms < 30000);

Why 30 seconds? could/should this be a parameter?

**Comment by flatherz** [comment]
@magicrub things went sideways while I working on rebasing my changes. I certainly didn't intend to force merge this PR with no changes! Working on reverting things back...

File: ArduPlane/Plane.cpp
Comment (magicrub) [reply ArduPlane/Plane.cpp]: meh. Doesn't need to be. I think after 30 seconds of flying you're no longer considered taking off. What number would you like?

**Comment by magicrub** [comment]
@flatherz I've made the changes that are needed to merge but you unchecked the box when making the PR allowing maintainers to add commits on your branch. So, please cherry-pick https://github.com/magicrub/ardupilot/commit/2ca4cde2eeab3cdbc25426d0b88a270965bbd64e from my repo branch located at https://github.com/magicrub/ardupilot/commits/pr-plane-is-taking-off

**Comment by flatherz** [comment]
> @flatherz I've made the changes that are needed to merge but you unchecked the box when making the PR allowing maintainers to add commits on your branch. So, please cherry-pick [magicrub@2ca4cde](https://github.com/magicrub/ardupilot/commit/2ca4cde2eeab3cdbc25426d0b88a270965bbd64e) from my repo branch located at https://github.com/magicrub/ardupilot/commits/pr-plane-is-taking-off

I've checked that box allowing maintainers to make edits, so it should work now. I think it might be safer if you make the changes you want seeing as how my attempt at rebasing went sideways.

**Comment by magicrub** [comment]
@flatherz done! Thanks for your work.

**Comment by magicrub** [comment]
Since I co-authored it, someone else should approve/merge it. Maybe @IamPete1 since he approved https://github.com/ArduPilot/ardupilot/pull/23827

File: ArduPlane/Plane.cpp
Comment (magicrub) [reply ArduPlane/Plane.cpp]: > > ```
> > return control_mode->is_taking_off() || (started_flying_ms > 0 && millis() - started_flying_ms < 30000);
> > ```
> 
> Why 30 seconds? could/should this be a parameter?

@timtuxworth If it was a parameter, what time duration would you use?

Comment (timtuxworth) [reply ArduPlane/Plane.cpp]: seconds. Default to 30 is fine, but depending on the size and capabilities of the vehicle or even the flying location, you might want it to be shorter or longer. Some might want longer to be absolutely sure it's really flying, some want to make it much shorter depending on what they are doing.

Comment (IamPete1) [reply ArduPlane/Plane.cpp]: I think I prefer the original patch. If were not sure from the flight mode specific handling then assume were not taking off. There is potential for `started_flying_ms` to be wrong.

File: ArduPlane/mode.h
Comment (IamPete1) [inline ArduPlane/mode.h:759]: I think this is too simplistic, need to check if takeoff has completed and the vehicles is loitering. Check for `AP_FixedWing::FlightStage::TAKEOFF`.

File: ArduPlane/mode_auto.cpp
Comment (IamPete1) [inline ArduPlane/mode_auto.cpp:176]: Again check for `AP_FixedWing::FlightStage::TAKEOFF` to see the the take off is done for fixedwing takeoff.

File: ArduPlane/mode_auto.cpp
Comment (IamPete1) [inline ArduPlane/mode_auto.cpp:168]: VTOL transition is not a landing.

File: ArduPlane/mode_auto.cpp
Comment (IamPete1) [inline ArduPlane/mode_auto.cpp:179]: This should not be here.

File: ArduPlane/mode_auto.cpp
Comment (IamPete1) [inline ArduPlane/mode_auto.cpp:166]: `AP_FixedWing::FlightStage::LAND`

File: ArduPlane/ArduPlane.cpp
Comment (IamPete1) [inline ArduPlane/ArduPlane.cpp:856]: Why `in_vtol_land_descent` and not `in_vtol_land_final` ? I'm not really sure which is the correct answer. Potentially land descent could start at a 100m altitude or or more.

**Review by IamPete1 — CHANGES_REQUESTED** [review]
There is quite a lot of nuance here that we will have to think about carefully.

File: ArduPlane/Plane.cpp
Comment (magicrub) [reply ArduPlane/Plane.cpp]: fair enough.. so just remove the 30s check for now? We can iterate on improvements later

File: ArduPlane/mode_auto.cpp
Comment (magicrub) [reply ArduPlane/mode_auto.cpp:179]: "this" meaning MAV_CMD_DO_VTOL_TRANSITION or the whole switch?

File: ArduPlane/ArduPlane.cpp
Comment (magicrub) [reply ArduPlane/ArduPlane.cpp:856]: I'd say either one and/or both are valid landing states. If I was a landing gear or an LED blinky status thing I'd want to know "landing" meant a land descent and not just the final moments

File: ArduPlane/Plane.cpp
Comment (magicrub) [reply ArduPlane/Plane.cpp]: I see in TAKEOFF mode it's hardcoded to switch from flight_stage "TAKEOFF" to "NORMAL after s(tarted_flying_ms > 10000 && groundspeed > 3)

File: ArduPlane/mode.h
Comment (magicrub) [reply ArduPlane/mode.h:759]: Flight_stage is "NORMAL" for all flight modes except TAKEOFF and AUTO.. unless VTOL is overriding it

File: ArduPlane/mode_auto.cpp
Comment (IamPete1) [reply ArduPlane/mode_auto.cpp:179]: Sorry, Just `MAV_CMD_DO_VTOL_TRANSITION `.

Comment (magicrub) [reply ArduPlane/mode_auto.cpp]: removed the whole switch

File: ArduPlane/mode_auto.cpp
Comment (magicrub) [reply ArduPlane/mode_auto.cpp]: removed the whole switch and using flight stage

File: ArduPlane/mode_auto.cpp
Comment (magicrub) [reply ArduPlane/mode_auto.cpp]: removed the whole switch and using flight stage

File: ArduPlane/mode_auto.cpp
Comment (magicrub) [reply ArduPlane/mode_auto.cpp]: changed to using flight stage

File: ArduPlane/mode.h
Comment (magicrub) [reply ArduPlane/mode.h]: changed to using flight stage in mode.cpp

File: ArduPlane/Plane.cpp
Comment (magicrub) [reply ArduPlane/Plane.cpp]: I've removed the timer, it's using Flight Stage now per @IamPete1 's suggestion

**Comment by magicrub** [comment]
@IamPete1 all good now?

File: ArduPlane/mode_auto.cpp
Comment (IamPete1) [inline ArduPlane/mode_auto.cpp:165]: I'm not sure we should use `get_in_landing_sequence_flag` that can be set for log parts of missions that may not even have a actual landing at the end. 

I would be tempted to just put `plane.flight_stage == AP_FixedWing::FlightStage::LAND` in the base class method and forget about the auto override.

Comment (magicrub) [reply ArduPlane/mode_auto.cpp:165]: I think it's worthwhile to keep in. Same applies to the q VTOL landing stuff. It's vague but it's showing intent

Comment (IamPete1) [reply ArduPlane/mode_auto.cpp:165]: At least the VTOL landing stuff will actually end in a landing eventually. `get_in_landing_sequence_flag` just means we have been past a `DO_LAND_START`. The VTOL landing stuff also lands at the current location, the lading sequence can start a long way away from where the actual landing is going to happen.

Comment (magicrub) [reply ArduPlane/mode_auto.cpp:165]: For the sake of getting the rest in, I'll remove the mission check but I'd like to discuss adding it on a follow up PR

Comment (magicrub) [reply ArduPlane/mode_auto.cpp]: removed, only the stage is used now

**Comment by magicrub** [comment]
@IamPete1 I labeled it merge on CI pass..

**Pull request merged**
